File: blind75/33_number_of_connected_components/connected_components.go
package connected_components

// CountComponents counts connected components using DFS
// Time: O(V + E), Space: O(V + E)
func CountComponents(n int, edges [][]int) int {
	// Build adjacency list
	graph := make([][]int, n)
	for _, edge := range edges {
		a, b := edge[0], edge[1]
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a)
	}

	visited := make([]bool, n)
	count := 0

	var dfs func(node int)
	dfs = func(node int) {
		visited[node] = true
		for _, neighbor := range graph[node] {
			if !visited[neighbor] {
				dfs(neighbor)
			}
		}
	}

	// Count components
	for i := 0; i < n; i++ {
		if !visited[i] {
			count++
			dfs(i)
		}
	}

	return count
}

// CountComponentsBFS counts connected components using BFS
// Time: O(V + E), Space: O(V + E)
func CountComponentsBFS(n int, edges [][]int) int {
	// Build adjacency list
	graph := make([][]int, n)
	for _, edge := range edges {
		a, b := edge[0], edge[1]
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a)
	}

	visited := make([]bool, n)
	count := 0

	bfs := func(start int) {
		queue := []int{start}
		visited[start] = true

		for len(queue) > 0 {
			node := queue[0]
			queue = queue[1:]

			for _, neighbor := range graph[node] {
				if !visited[neighbor] {
					visited[neighbor] = true
					queue = append(queue, neighbor)
				}
			}
		}
	}

	// Count components
	for i := 0; i < n; i++ {
		if !visited[i] {
			count++
			bfs(i)
		}
	}

	return count
}

// UnionFind data structure
type UnionFind struct {
	parent     []int
	rank       []int
	components int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
	}
	return &UnionFind{parent: parent, rank: rank, components: n}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // Path compression
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) {
	rootX := uf.Find(x)
	rootY := uf.Find(y)

	if rootX == rootY {
		return // Already in same component
	}

	// Union by rank
	if uf.rank[rootX] < uf.rank[rootY] {
		uf.parent[rootX] = rootY
	} else if uf.rank[rootX] > uf.rank[rootY] {
		uf.parent[rootY] = rootX
	} else {
		uf.parent[rootY] = rootX
		uf.rank[rootX]++
	}

	uf.components--
}

// CountComponentsUnionFind counts connected components using Union Find
// Time: O(V + E × α(V)), Space: O(V)
func CountComponentsUnionFind(n int, edges [][]int) int {
	uf := NewUnionFind(n)

	for _, edge := range edges {
		a, b := edge[0], edge[1]
		uf.Union(a, b)
	}

	return uf.components
}

File: blind75/33_number_of_connected_components/connected_components.py
from typing import List
from collections import defaultdict, deque


def count_components(n: int, edges: List[List[int]]) -> int:
    """
    Counts connected components using DFS.
    Time: O(V + E), Space: O(V + E)

    Args:
        n: Number of nodes
        edges: List of edges [a, b]

    Returns:
        Number of connected components
    """
    # Build adjacency list
    graph = defaultdict(list)
    for a, b in edges:
        graph[a].append(b)
        graph[b].append(a)

    visited = set()
    count = 0

    def dfs(node: int):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)

    # Count components
    for i in range(n):
        if i not in visited:
            count += 1
            dfs(i)

    return count


def count_components_bfs(n: int, edges: List[List[int]]) -> int:
    """
    Counts connected components using BFS.
    Time: O(V + E), Space: O(V + E)

    Args:
        n: Number of nodes
        edges: List of edges [a, b]

    Returns:
        Number of connected components
    """
    # Build adjacency list
    graph = defaultdict(list)
    for a, b in edges:
        graph[a].append(b)
        graph[b].append(a)

    visited = set()
    count = 0

    def bfs(start: int):
        queue = deque([start])
        visited.add(start)

        while queue:
            node = queue.popleft()
            for neighbor in graph[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

    # Count components
    for i in range(n):
        if i not in visited:
            count += 1
            bfs(i)

    return count


class UnionFind:
    """Union Find data structure with path compression and union by rank."""

    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n
        self.components = n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x: int, y: int):
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return  # Already in same component

        # Union by rank
        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

        self.components -= 1


def count_components_union_find(n: int, edges: List[List[int]]) -> int:
    """
    Counts connected components using Union Find.
    Time: O(V + E × α(V)), Space: O(V)

    Args:
        n: Number of nodes
        edges: List of edges [a, b]

    Returns:
        Number of connected components
    """
    uf = UnionFind(n)

    for a, b in edges:
        uf.union(a, b)

    return uf.components
